test/test_typing.py

```diff
@@ -0,0 +1,298 @@
+# Copyright 2023-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test that each file in mypy_fails/ actually fails mypy, and test some
+sample client code that uses Motor typings.
+"""
+import os
+import sys
+import unittest
+from test.asyncio_tests import AsyncIOTestCase, asyncio_test
+from typing import TYPE_CHECKING, Any, AsyncIterable, Dict, Iterable, List, Union
+
+from bson import CodecOptions
+from bson.raw_bson import RawBSONDocument
+from bson.son import SON
+from pymongo.operations import DeleteOne, InsertOne, ReplaceOne
+from pymongo.read_preferences import ReadPreference
+
+from motor.core import AgnosticClient, AgnosticCollection
+
+try:
+    from bson import ObjectId
+    from typing_extensions import NotRequired, TypedDict
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+
+    class MovieWithId(TypedDict):
+        _id: ObjectId
+        name: str
+        year: int
+
+    class ImplicitMovie(TypedDict):
+        _id: NotRequired[ObjectId]
+        name: str
+        year: int
+
+except ImportError:
+    Movie = dict  # type:ignore[misc,assignment]
+    ImplicitMovie = dict  # type: ignore[assignment,misc]
+    MovieWithId = dict  # type: ignore[assignment,misc]
+    TypedDict = None
+    NotRequired = None  # type: ignore[assignment]
+
+
+try:
+    from mypy import api
+except ImportError:
+    api = None
+
+sys.path[0:0] = [""]
+
+
+TEST_PATH = os.path.join(os.path.dirname(os.path.realpath(__file__)), "mypy_fails")
+
+
+def get_tests() -> Iterable[str]:
+    for dirpath, _, filenames in os.walk(TEST_PATH):
+        for filename in filenames:
+            yield os.path.join(dirpath, filename)
+
+
+def only_type_check(func):
+    def inner(*args, **kwargs):
+        if not TYPE_CHECKING:
+            raise unittest.SkipTest("Used for Type Checking Only")
+        func(*args, **kwargs)
+
+    return inner
+
+
+class TestMypyFails(unittest.TestCase):
+    def ensure_mypy_fails(self, filename: str) -> None:
+        if api is None:
+            raise unittest.SkipTest("Mypy is not installed")
```

Comment:
We need to run this test (or the entire test suite) with mypy installed. Right now I believe these tests are always skipped.